pkg/plugin/validator.go

```diff
@@ -0,0 +1,189 @@
+)
+
+// Validator validates github issue against validation criteria.
+type Validator struct {
+	cfg        *PluginConfig
+	decodedPem *rsa.PrivateKey
+	client     *github.Client
+	githubApp  *githubapp.GitHubApp
+}
+
+// ExchangeResponse is the GitHub API response of requesting an access token
+// for the GitHub App installation with requested repositories and permissions.
+type ExchangeResponse struct {
+	AccessToken string `json:"token"`
+}
+
+// pluginGithubIssue contains the required attribute parsed from
+// the issue URL.
+type pluginGithubIssue struct {
+	Owner       string
+	RepoName    string
+	IssueNumber int
+}
+
+type ValidatorOption func(*Validator)
+
+func WithGitHubClient(c *github.Client) ValidatorOption {
+	return func(v *Validator) {
+		v.client = c
+	}
+}
+
+func WithGithubApp(c *githubapp.GitHubApp) ValidatorOption {
+	return func(v *Validator) {
+		v.githubApp = c
+	}
+}
+
+// NewValidator creates a validator.
+func NewValidator(cfg *PluginConfig, opts ...ValidatorOption) (*Validator, error) {
+	pk, err := readPrivateKey(cfg.GitHubAppPrivateKeyPEM)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read private key: %w", err)
+	}
+	v := Validator{
+		cfg:        cfg,
+		decodedPem: pk,
+	}
+	for _, opt := range opts {
+		opt(&v)
+	}
+	if v.client == nil {
+		v.client = github.NewClient(nil)
+	}
+	if v.githubApp == nil {
+		ghCfg := githubapp.NewConfig(v.cfg.GitHubAppID, v.cfg.GitHubAppInstallationID, v.decodedPem, githubapp.WithJWTTokenCaching(1*time.Minute))
+		v.githubApp = githubapp.New(ghCfg)
+	}
+	return &v, nil
+}
+
+// MatchIssue parses issue info from provided issueURL and validate if the issue is valid.
+func (v *Validator) MatchIssue(ctx context.Context, issueURL string, opts ...githubapp.ConfigOption) error {
+	info, err := parseIssueInfoFromURL(issueURL)
+	if err != nil {
+		return fmt.Errorf("failed to parse issueURL: %w", err)
+	}
+
+	t, err := v.getAccessToken(ctx, info.RepoName)
+	if err != nil {
+		return fmt.Errorf("failed to get access token: %w", err)
+	}
+	v.client = v.client.WithAuthToken(t)
+
+	return v.validateIssue(ctx, info)
+}
+
+// validateIssue verifies if the issue exists and the issue is open.
+func (v *Validator) validateIssue(ctx context.Context, pi *pluginGithubIssue) error {
+	issue, err := v.getGithubIssue(ctx, pi)
+	if err != nil {
+		return fmt.Errorf("failed to get issue info: %w", err)
+	}
+	if s := issue.GetState(); s != "open" {
+		return fmt.Errorf("issue is in state: %s, please make sure to use an open issue", s)
+	}
+	return nil
+}
+
+// getGithubIssue gets the provided issue's info from github api.
+func (v *Validator) getGithubIssue(ctx context.Context, pi *pluginGithubIssue) (*github.Issue, error) {
+	issue, _, err := v.client.Issues.Get(ctx, pi.Owner, pi.RepoName, pi.IssueNumber)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get issue: %w", err)
+	}
+	return issue, nil
+}
```

Comment:
nit: this func seems pretty useless. It's only used in one place and there is no net gain (you actually introduced 6 more lines of code with some useless error nesting)

Reply:
removed.